Wrappers/Python/cil/optimisation/functions/SAGFunction.py

```diff
@@ -0,0 +1,167 @@
+        """ Returns the gradient of the selected function or batch of functions at :code:`x`. 
+            The function_num is selected using the sampler. 
+
+        Parameters
+        ----------
+        x: element in the domain of the `functions`
+
+        function_num: `int` 
+            Between 0 and the number of functions in the list  
+
+        """
+        if self._list_stored_gradients is None: 
+            self._list_stored_gradients = [
+                0*x for fi in self.functions]
+            self._full_gradient_at_iterate = 0*x
+            
+
+        self._stoch_grad_at_iterate = self.functions[function_num].gradient(x)
+
+        
+        self._update_data_passes_indices([self.function_num])
+
+        self._stochastic_grad_difference = self._stoch_grad_at_iterate.sapyb(
+            1., self._list_stored_gradients[function_num], -1.)
+
+   
+        out =self._update_approx_gradient(out)
+
+        self._list_stored_gradients[function_num].fill(
+            self._stoch_grad_at_iterate)
+        self._full_gradient_at_iterate.sapyb(
+            1., self._stochastic_grad_difference, 1., out=self._full_gradient_at_iterate)
+
+        return out
+    
+    def _update_approx_gradient(self, out):
+        """Internal function used to differentiate between the SAG and SAGA calculations"""
+        if out is None:
+            # due to the convention that we follow: without the 1/n factor
+            out = self._stochastic_grad_difference.sapyb(
+                1., self._full_gradient_at_iterate, 1.)
+        else:
+            # due to the convention that we follow: without the 1/n factor
+            self._stochastic_grad_difference.sapyb(
+                1., self._full_gradient_at_iterate, 1., out=out)
+
+        return out 
+    
+    def warm_start_approximate_gradients(self, initial):
+        """A function to warm start SAG or SAGA algorithms by initialising all the gradients at an initial point.
+        
+        Parameters
+        ----------
+        initial: DataContainer,
+            The initial point to warmstart the calculation 
+        
+        """
+        self._list_stored_gradients = [
+            fi.gradient(initial) for fi in self.functions]
+        self._full_gradient_at_iterate = np.sum(self._list_stored_gradients)
+        self._update_data_passes_indices(list(range(self.num_functions)))
+
+    @property
+    def data_passes_indices(self):
+        ret = self._data_passes_indices[:]  
+        if len(ret[0]) == self.num_functions:  
+            a = ret.pop(1)  
+            ret[0] += a  
+        return ret
+    
+class SAGAFunction(SAGFunction):
+
+    """
+    An accelerated version of the stochastic average gradient function, a child class of `ApproximateGradientSumFunction`, which defines from a list of functions, :math:`{f_1,...,f_n}` a `SumFunction`, :math:`f_1+...+f_n` where each time the `gradient` is called, the `sampler` provides an index, :math:`i \in {1,...,n}` 
```

Comment:
Please clarify how it differs from SAG. What is the acceleration? (maybe I missed it).

Reply:
I can't see that this has been addressed? I think a sentence to explain how SAGA is different from SAG, what the acceleration is, would be suitable.

Reply:
Thanks you for elaborating this!